mne/preprocessing/ica.py

```diff
@@ -640,11 +661,54 @@ def _pre_whiten(self, data, info, picks):
     def _fit(self, data, max_pca_components, fit_type):
         """Aux function."""
         random_state = check_random_state(self.random_state)
-        pca = _PCA(n_components=max_pca_components, whiten=True)
         n_channels, n_samples = data.shape
-        data = pca.fit_transform(data.T)
-        assert data.shape == (n_samples, max_pca_components or n_channels)
 
+        if max_pca_components is None or isinstance(max_pca_components, float):
+            # Use all channels. For float input, we will reduce the number
+            # of retained components later.
+            self.max_pca_components_ = n_channels
+        else:
+            self.max_pca_components_ = max_pca_components
+
+        pca = _PCA(n_components=self.max_pca_components_, whiten=True)
+        data_transformed = pca.fit_transform(data.T)
+
+        # If user passed a float, re-run the PCA, but only keep the number of
+        # components required to explain the requested cumulative variance.
+        # While we could also simply select a sub-set of the data acquired
+        # above, re-running PCA from scratch is less bug-prone, as it will
+        # ensure that all internals of the _PCA instance are properly updated
+        # as well, and we don't need to handle this case specially below.
+        # Note that the result should be numerically identical to the first
+        # PCA run for all retained components.
+        if isinstance(max_pca_components, float):
+            del data_transformed  # Free memory.
+            self.max_pca_components_ = (np.sum(pca.explained_variance_ratio_
+                                               .cumsum()
+                                        <= max_pca_components))
+
+            if (self.n_components is not None and
+                    self.max_pca_components_ < self.n_components):
+                msg = (f'You asked to select only a subset of PCA components '
```

Comment:
Minor nitpick, this actually requires more left spaces than the pattern
```
raise ValueError(
    f''
    f'')
```
and it also makes it look like `msg` could be triaged later (in a `on_whatever='raise'` sort of pattern) / you have to look down more lines to know all this conditional does is raise an error

Reply:
good point, will keep that in mind for future PRs